### PR TITLE
feat(normalize_features): Added function and tests

### PR DESCRIPTION
**Pull request by ruiheesi**
This function is used for normalizing each column of the dataset with the low and high quantile for each role.

### CODE REVIEW

**Comment by ruiheesi** [comment]
Hi George, this is updated quantile normalization with different low and high quantile as discussed. Thanks.

File: src/spac/transformations.py [hunk 1]
```diff
@@ -261,3 +262,137 @@ def rename_observations(adata, src_observation, dest_observation, mappings):
...
+    adata: anndata,
+    low_quantile: float = 0.02,
+    high_quantile: float = 0.02,
+    target_layer: str = None,
+    new_layer_name: str = "normalized_feature",
+    overwrite: bool = True
+):
+
+    """
+    Normalized the features of a dataset stored in an AnnData object.
```
Comment (georgezakinih) [inline src/spac/transformations.py:277]: GZ: Normalize the features  stored in an AnnData object.

File: src/spac/transformations.py [hunk 2]
```diff
@@ -261,3 +262,137 @@ def rename_observations(adata, src_observation, dest_observation, mappings):
         )
 
     return adata
+
+
+def normalize_features(
+    adata: anndata,
+    low_quantile: float = 0.02,
+    high_quantile: float = 0.02,
+    target_layer: str = None,
```
Comment (georgezakinih) [inline src/spac/transformations.py:271]: Can we use "input_layer" instead of "target_layer"?

File: src/spac/transformations.py [hunk 3]
```diff
@@ -261,3 +262,137 @@ def rename_observations(adata, src_observation, dest_observation, mappings):
...
+    Parameters
+    ----------
+    adata : anndata.AnnData
+        An AnnData object containing the data to be normalized.
+
+    low_quantile : float, optional (default: 0.02)
+        The lower quantile to use for normalization. Determines the
+        minimum value after normalization. Must be a positive float.
+
+    high_quantile : float, optional (default: 0.02)
```
Comment (georgezakinih) [inline src/spac/transformations.py:293]: must be positive float [0,1[

File: src/spac/transformations.py [hunk 4]
```diff
@@ -261,3 +262,137 @@ def rename_observations(adata, src_observation, dest_observation, mappings):
...
+        The higher quantile to use for normalization. Determines the
```
Comment (georgezakinih) [inline src/spac/transformations.py:294]: positive float ]0,1]

File: src/spac/transformations.py [hunk 5]
```diff
@@ -261,3 +262,137 @@ def rename_observations(adata, src_observation, dest_observation, mappings):
...
+        maximum value after normalization. Must be a positive float.
+
+    target_layer : str, optional (default: None)
+        The name of the layer in the AnnData object to be normalized.
+        If None, the function will use the raw data layer.
```
Comment (georgezakinih) [inline src/spac/transformations.py:299]: "default" instead of "raw" data layer

File: src/spac/transformations.py [hunk 6]
```diff
@@ -261,3 +262,137 @@ def rename_observations(adata, src_observation, dest_observation, mappings):
...
+                         "the high quantile, currently value is:\n"
+                         f"low quantile: {low_quantile}\n"
+                         f"high quantile: {high_quantile}")
+
+    dataframe = adata.to_df(layer=target_layer)
+
+    # Calculate low and high quantiles
+    quantiles = dataframe.quantile([low_quantile, high_quantile])
+
+    new_row_names = {
```
Comment (georgezakinih) [inline src/spac/transformations.py:350]: What is new_row_name?

File: src/spac/transformations.py [hunk 7]
```diff
@@ -261,3 +262,137 @@ def rename_observations(adata, src_observation, dest_observation, mappings):
...
+        high_quantile: 'quantile_high',
+        low_quantile: 'quantile_low'
+    }
+
+    # reassign the row names for downstream process
+
+    quantiles.index = quantiles.index.map(new_row_names)
+
+    pre_info = dataframe.describe()
```
Comment (georgezakinih) [inline src/spac/transformations.py:359]: This variable pre_info is overwritten in the following line. Do we need this?

File: src/spac/transformations.py [hunk 8]
```diff
@@ -261,3 +262,137 @@ def rename_observations(adata, src_observation, dest_observation, mappings):
...
+        If the new layer name exists in the anndata object,
+        the function will defaultly overwrite the existing table unless
+        'overwrite' is False
+
+    Returns
+    -------
+    None
+        This function directly modifies the input AnnData object in-place
+        by adding the new scaled layer, and stores related information in
+        anndata.uns.
```
Comment (georgezakinih) [inline src/spac/transformations.py:315]: Would you please elaborate on what is stored in anndata.uns? Maybe add an example showing what gets printed?

File: src/spac/utils.py [hunk 9]
```diff
@@ -70,3 +71,229 @@ def regex_search(pattern, str):
     all_results = list(all_results)
 
     return all_results
+
+
+def check_list_in_list(
```
Comment (georgezakinih) [inline src/spac/utils.py:76]: Just curious why this is part of this commit? I thought we already merged this part.

File: tests/test_transformations/test_normalize_features.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        # Create a sample AnnData object with a test dataframe
+        self.data = {
+            'feature1': [0.1, 0.2, 0.3, 0.4, 0.5],
+            'feature2': [0.6, 0.7, 0.8, 0.9, 1.0]
+        }
+        self.adata = AnnData(pd.DataFrame(self.data))
+
+    def test_invalid_high_quantile(self):
+        # Test with an invalid high_quantile (not numeric)
+        with self.assertRaises(TypeError):
```
Comment (georgezakinih) [inline tests/test_transformations/test_normalize_features.py:19]: What about we test coroner case, so you want to use something like 1.01?

File: tests/test_transformations/test_normalize_features.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+            normalize_features(self.adata, high_quantile='2')
+
+    def test_invalid_low_quantile(self):
+        # Test with an invalid low_quantile (not numeric)
+        with self.assertRaises(TypeError):
+            normalize_features(self.adata, low_quantile='2')
+
+    def test_high_quantile_out_of_range(self):
+        # Test with high_quantile out of range (should be within (0, 1])
```
Comment (georgezakinih) [inline tests/test_transformations/test_normalize_features.py:28]: Can we test with the corner case 1.01

File: tests/test_transformations/test_normalize_features.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        with self.assertRaises(ValueError):
+            normalize_features(self.adata, high_quantile=1.5)
+
+    def test_low_quantile_out_of_range(self):
```
Comment (georgezakinih) [inline tests/test_transformations/test_normalize_features.py:32]: can we test with the corner case -0.01

File: tests/test_transformations/test_normalize_features.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    def test_low_quantile_greater_than_high_quantile(self):
+        # Test with low_quantile greater than high_quantile
+        with self.assertRaises(ValueError):
+            normalize_features(
+                self.adata,
+                low_quantile=0.8,
+                high_quantile=0.7
+            )
+
+    def test_correct_output(self):
```
Comment (georgezakinih) [inline tests/test_transformations/test_normalize_features.py:46]: I suggest to use a scenario that is easy to reason about, e.g, inputs are from 0 to 1, or 100 to 200. This way a developer can know the "expected output" right away without calculating it.

**Review by georgezakinih — COMMENTED** [review]
Thanks Rui. I added few comments. Would you please review and let me know what you think?

File: src/spac/transformations.py (hunk 4 above)
Comment (ruiheesi) [reply src/spac/transformations.py:294]: The high quantile should be (0,1] as low can be [0, 1), high must be greater than low

File: src/spac/utils.py (hunk 9 above)
Comment (ruiheesi) [reply src/spac/utils.py:76]: Ahh sorry the merge should be to dev not main, I will change this. Once it is to dev, this should not appear.
